### DISCUSSION TITLE
How to use with CFFI bindings?

### DISCUSSION THREAD

**Discussion by yaqwsx** [messense/maturin-action#147] [Q&A] [answered]
Hello, I am trying to use maturin-action and Maturin for building and packaging a Rust library that exposes CFFI binding. Meanwhile everything works fine locally, the GH action using maturin-action fails.

It fails on:
```
Caused by: Failed to generate cffi declarations using /opt/python/cp36-cp36m/bin/python3: exit status: 1
Traceback (most recent call last):
  File "<string>", line 2, in <module>
ModuleNotFoundError: No module named 'cffi'
```

CFFI is specified as build dependency. Is it the correct place to specify it?

This is my [`pyproject.toml`](https://github.com/yaqwsx/py_cavalier_contours/blob/main/pyproject.toml) and [`Cargo.toml`](https://github.com/yaqwsx/py_cavalier_contours/blob/main/Cargo.toml) and [CI](https://github.com/yaqwsx/py_cavalier_contours/blob/main/.github/workflows/CI.yml).

**Comment by messense** [comment, marked as answer]
`cffi` support in `maturin-action` is rather unpolished ATM, you need to install `cffi` manually, `maturin` only tries to install it when using a virtualenv.

**Reply by yaqwsx** [reply]
Thank you for your reply. Where should I put the installation? E.g., the Linux version runs inside the manylinux container, which is created and managed by the action.

**Reply by messense** [reply]
I'm going to pre-install `cffi` in `manylinux-cross` docker image: https://github.com/rust-cross/manylinux-cross/pull/34

**Reply by messense** [reply]
Also added `cffi` installation to `maturin-action` in https://github.com/PyO3/maturin-action/pull/150, please try again.

**Reply by yaqwsx** [reply]
Thanks, it works now!
